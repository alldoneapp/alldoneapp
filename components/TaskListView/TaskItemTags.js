import React, { useEffect, useState } from 'react'
import { Dimensions, StyleSheet, View } from 'react-native'
import TaskSummarizeTags from '../Tags/TaskSummarizeTags'
import { useSelector } from 'react-redux'
import { checkIfInMyDay } from '../MyDayView/MyDayTasks/MyDayOpenTasks/myDayOpenTasksHelper'
import Tags from './TagsArea/Tags'
import useTagsAmount from './TagsArea/useTagsAmount'
import ProjectTag from '../Tags/ProjectTag'

const TaskItemTags = ({
    task,
    isSubtask,
    projectId,
    isObservedTask,
    isToReviewTask,
    toggleSubTaskList,
    subtaskList,
    isActiveOrganizeMode,
    accessGranted,
    anonymousGranted,
    forceTagsMobile,
    setTagsExpandedHeight,
    isLocked,
    isSuggested,
    isPending,
    setExpandTags,
    expandTags,
    showSummarizeTagInByTime,
}) => {
    const showAllProjectsByTime = useSelector(state => state.loggedUser.showAllProjectsByTime)
    const route = useSelector(state => state.route)
    const selectedSidebarTab = useSelector(state => state.selectedSidebarTab)
    const taskViewToggleIndex = useSelector(state => state.taskViewToggleIndex)
    const selectedProjectIndex = useSelector(state => state.selectedProjectIndex)
    const tablet = useSelector(state => state.isMiddleScreen)
    const smallScreenNavigation = useSelector(state => state.smallScreenNavigation)
    const smallScreenNavSidebarCollapsed = useSelector(state => state.smallScreenNavSidebarCollapsed)
    const [visible, setVisible] = useState(false)

    const amountTags = useTagsAmount({
        task,
        isSubtask,
        projectId,
        isObservedTask,
        isToReviewTask,
        subtaskList,
        isSuggested,
        isPending,
    })

    const inMyDayAndNotSubtask =
        checkIfInMyDay(selectedProjectIndex, showAllProjectsByTime, route, selectedSidebarTab, taskViewToggleIndex) &&
        !isSubtask

    const isMobile = smallScreenNavigation || smallScreenNavSidebarCollapsed

    const needSummarize = inMyDayAndNotSubtask
        ? amountTags > 0 && showSummarizeTagInByTime
        : amountTags > 5 || (tablet && amountTags > 3) || (isMobile && amountTags > 2)

    const toggleVisibleTags = e => {
        e.preventDefault()
        e.stopPropagation()
        setTagsExpanded(!visible)
        setVisible(!visible)
        setExpandTags?.(!expandTags)
    }

    const setTagsExpanded = isVisible => {
        if (isVisible && needSummarize) {
            setTagsExpandedHeight(isSubtask ? 32 : 36)
        } else {
            setTagsExpandedHeight(0)
        }
    }

    useEffect(() => {
        setTagsExpanded(visible)
    }, [tablet, isMobile])

    useEffect(() => {
        if (!needSummarize) {
            setTagsExpandedHeight(0)
            setExpandTags?.(false)
        }
    }, [needSummarize])

    return (
        <>
            {needSummarize && (
                <View
                    style={[
                        localStyles.container,
                        visible &&
                            !inMyDayAndNotSubtask &&
                            (isSubtask ? localStyles.containerExpandSubTasks : localStyles.containerExpand),
                    ]}
                >
                    <View style={{ flexDirection: 'row' }}>
                        <TaskSummarizeTags amountTags={amountTags} onPress={toggleVisibleTags} />
                        {inMyDayAndNotSubtask && (
                            <ProjectTag
                                style={{ marginLeft: 8 }}
                                projectId={projectId}
                                disabled={isActiveOrganizeMode || isLocked || !accessGranted}
                                shrinkTextToAmountOfLetter={8}
                            />
                        )}
                    </View>
                    {visible && !inMyDayAndNotSubtask && (
                        <View style={[localStyles.tagsGroup, isSubtask && localStyles.tagsGroupSubTasks]}>
                            <Tags
                                task={task}
                                isSubtask={isSubtask}
                                projectId={projectId}
                                isObservedTask={isObservedTask}
                                isToReviewTask={isToReviewTask}
                                toggleSubTaskList={toggleSubTaskList}
                                subtaskList={subtaskList}
                                isActiveOrganizeMode={isActiveOrganizeMode}
                                accessGranted={accessGranted}
                                anonymousGranted={anonymousGranted}
                                forceTagsMobile={forceTagsMobile}
                                isLocked={isLocked}
                                isSuggested={isSuggested}
                                isPending={isPending}
                                needSummarize={needSummarize}
                            />
                        </View>
                    )}
                </View>
            )}
            {!needSummarize && (
                <Tags
                    task={task}
                    isSubtask={isSubtask}
                    projectId={projectId}
                    isObservedTask={isObservedTask}
                    isToReviewTask={isToReviewTask}
                    toggleSubTaskList={toggleSubTaskList}
                    subtaskList={subtaskList}
                    isActiveOrganizeMode={isActiveOrganizeMode}
                    accessGranted={accessGranted}
                    anonymousGranted={anonymousGranted}
                    forceTagsMobile={forceTagsMobile}
                    isLocked={isLocked}
                    isSuggested={isSuggested}
                    isPending={isPending}
                />
            )}
        </>
    )
}

const localStyles = StyleSheet.create({
    container: {
        flexDirection: 'column',
        flexGrow: 0,
        paddingLeft: 8,
    },
    containerExpand: {
        paddingBottom: 36,
    },
    containerExpandSubTasks: {
        paddingBottom: 32,
    },
    tagsGroup: {
        position: 'absolute',
        right: 0,
        width: Dimensions.get('screen').width - 24,
        alignContent: 'flex-end',
        justifyContent: 'flex-end',
        flexDirection: 'row',
        top: 36,
    },
    tagsGroupSubTasks: {
        top: 32,
    },
})

export default TaskItemTags
