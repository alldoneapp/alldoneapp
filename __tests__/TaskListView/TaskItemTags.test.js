/**
 * @jest-environment jsdom
 */

import React from 'react'
import renderer from 'react-test-renderer'
import TaskItemTags from '../../components/TaskListView/TaskItemTags'
import { Text } from 'react-native'

describe('TaskItemTags component', () => {
    describe('TaskItemTags snapshot test', () => {
        it('should render correctly', async () => {
            const tree = renderer
                .create(
                    <TaskItemTags amountTags={5}>
                        <Text>Some text</Text>
                    </TaskItemTags>
                )
                .toJSON()
            expect(tree).toMatchSnapshot()
        })
    })

    describe('TaskItemTags functions snapshot test', () => {
        it('check updateState function', () => {
            const tree = renderer.create(
                <TaskItemTags amountTags={0}>
                    <Text>Some text</Text>
                </TaskItemTags>
            )
            expect(tree.toJSON()).toMatchSnapshot()

            tree.getInstance().updateState()
            expect(tree.toJSON()).toMatchSnapshot()
        })

        it('check toggleVisibleTags function', () => {
            const tree = renderer.create(
                <TaskItemTags amountTags={1}>
                    <Text>Some text</Text>
                </TaskItemTags>
            )
            expect(tree.toJSON()).toMatchSnapshot()
            tree.getInstance().state.visible = false

            tree.getInstance().toggleVisibleTags()
            let state = tree.getInstance().state
            expect(state.visible).toBeTruthy()
        })

        it('check the unmount action', () => {
            const tree = renderer.create(
                <TaskItemTags amountTags={3}>
                    <Text>Some text</Text>
                </TaskItemTags>
            )
            tree.unmount()
        })
    })
})
