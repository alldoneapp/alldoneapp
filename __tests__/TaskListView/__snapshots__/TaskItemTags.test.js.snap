// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskItemTags component TaskItemTags functions snapshot test check toggleVisibleTags function 1`] = `
<Text>
  Some text
</Text>
`;

exports[`TaskItemTags component TaskItemTags functions snapshot test check updateState function 1`] = `
<Text>
  Some text
</Text>
`;

exports[`TaskItemTags component TaskItemTags functions snapshot test check updateState function 2`] = `
<Text>
  Some text
</Text>
`;

exports[`TaskItemTags component TaskItemTags snapshot test should render correctly 1`] = `
<Text>
  Some text
</Text>
`;
